pennylane/ops/PauliArithmetic.py

```diff
@@ -0,0 +1,194 @@
+# Copyright 2018-2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from enum import Enum
+from functools import reduce
+
+from pennylane import math
+import numpy as np
+from scipy import sparse
+
+
+class Pauli(Enum):
+    I = 0
+    X = 1
+    Y = 2
+    Z = 3
+
+    def __lt__(self, other):
+        return self.value < other.value
+
+    def __repr__(self):
+        return self.name
+
+
+I = Pauli.I
+X = Pauli.X
+Y = Pauli.Y
+Z = Pauli.Z
+
+matX = np.array([[0, 1], [1, 0]])
+matY = np.array([[0, -1j], [1j, 0]])
+matZ = np.array([[1, 0], [0, -1]])
+matI = np.eye(2)
+
+mat_map = {
+    I: matI,
+    X: matX,
+    Y: matY,
+    Z: matZ,
+}
+
+sparse_matX = sparse.csr_matrix([[0, 1], [1, 0]])
+sparse_matY = sparse.csr_matrix([[0, -1j], [1j, 0]])
+sparse_matZ = sparse.csr_matrix([[1, 0], [0, -1]])
+sparse_matI = sparse.eye(2, format="csr")
+
+sparse_mat_map = {
+    I: sparse_matI,
+    X: sparse_matX,
+    Y: sparse_matY,
+    Z: sparse_matZ,
+}
+
+
+class PauliWord(dict):
+    """Immutable dictionary used to represent a Pauli Word.
+    Can be constructed from a standard dictionary.
+
+    >>> w = PauliWord({"a": X, 2: Y, 3: z})
+    """
+    _map_I = {
+        I: (1, I),
+        X: (1, X),
+        Y: (1, Y),
+        Z: (1, Z),
+    }
+    _map_X = {
+        I: (1, X),
+        X: (1, I),
+        Y: (1.0j, Z),
+        Z: (-1.0j, Y),
+    }
+    _map_Y = {
+        I: (1, Y),
+        X: (-1.0j, Z),
+        Y: (1, I),
+        Z: (1j, X),
+    }
+    _map_Z = {
+        I: (1, Z),
+        X: (1j, Y),
+        Y: (-1.0j, X),
+        Z: (1, I),
+    }
+
+    mul_map = {
+        I: _map_I,
+        X: _map_X,
+        Y: _map_Y,
+        Z: _map_Z
+    }
```

Comment:
Not sure if this dictionary needs to belong to the class. We just need to have the class in this namespace where the `PauliWord` class is defined.

Reply:
Moved out 👍🏼